internal/stacks/stackruntime/internal/stackeval/component_instance.go

```diff
@@ -126,8 +135,67 @@ func (c *ComponentInstance) inputValuesForModulesRuntime(ctx context.Context, ph
 			Value:      v,
 			SourceType: terraform.ValueFromCaller,
 		}
+
+		// While we're here, we'll just add a diagnostic if the value has
+		// somehow changed between the planning and apply phases. All of these
+		// diagnostics acknowledge that the root cause here is a bug in
+		// Terraform.
+		if phase == ApplyPhase {
+			raw, ok := previousValues[name]
+			if !ok {
+				// This shouldn't happen because we should have a value for
+				// every input variable that we have a value for in the plan.
+				// TODO: Support for ephemeral values is incoming, once that
+				//   is implemented it will be possible for there to be a
+				//   different set of input variables between the plan and the
+				//   apply phase.
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Missing input variable value",
+					Detail: fmt.Sprintf(
+						"The input variable %q is required but was not set in the plan for %s. This is a bug in Terraform - please report it.",
+						name, c.Addr(),
+					),
+					Subject: c.call.Declaration(ctx).DeclRange.ToHCL().Ptr(),
+				})
+				continue
+			}
+
+			plannedValue, err := raw.Decode(cty.DynamicPseudoType)
+			if err != nil {
+				// Then something has gone wrong when decoding the value.
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid planned input variable value",
+					Detail:   fmt.Sprintf("Failed to decode the planned value for input variable %q: %s. This is a bug in Terraform - please report it.", name, err),
+					Subject:  c.call.Declaration(ctx).DeclRange.ToHCL().Ptr(),
+				})
+				continue
+			}
+
+			if equals, _ := plannedValue.Equals(v).Unmark(); !equals.IsKnown() {
+				// We unmark the value as we don't care about the actual value,
+				// only whether it was equal or not.
+				//
+				// An unknown equals value means that the value was unknown
+				// during the planning stage so we'll just accept the apply
+				// value and not raise any diagnostics.
+			} else if !equals.True() {
```

Comment:
I think this particular case ought to be skipped if the input variable is declared as being ephemeral, since those _are_ allowed to change between plan and apply.

But I guess that's easier said than done because this function is doing its work in terms of the derived `wantTy` -- a `cty.Type` representing an object type -- instead of the underlying `map[string]*configs.Variable` it was derived from. :thinking:

Reply:
Originally I had imagined that ephemeral values wouldn't get here since they'd be missing from the `previousValues` map (as they wouldn't be stored in the plan).

I've updated it now so that they still won't even get here, but instead it does check the root module inputs and skips this check for any that are marked as ephemeral.